### PR TITLE
Add sign up and forgot password views

### PR DESCRIPTION
**Pull request by lewfish**
To test:

Try logging in with invalid username/password. The forgot link should pop up. Enter email address. Look at the console for the email. It isn't sent for real because we are using develop settings in Django. Follow the activation link. 

Try signing up for a new account. Follow the link in the console.

Try using input that has errors caught on the front-end (not filling required fields) and back-end errors (eg. passwords don't match, username is already in use, etc.).

I've noticed that the analyze view is unable to get the data to make the bar charts. I don't think this commit introduced the error. I might just need to run some secret command to get it to work (I tried destroying the VMs).

See https://github.com/WikiWatershed/model-my-watershed/issues/223

Connects #152

### CODE REVIEW

**Comment by caseypt** [comment]
Re: the analyze page not getting the data, that is due to the URL not having a trailing slash. I fixed it in the issue I'm working on.

**Comment by caseypt** [comment]
The form allows me to submit to the server without filling in the email field. I do get an "Email is invalid or already in use" message when the server responds, but we should prevent submitting the form if any of the fields are blank.

**Comment by caseypt** [comment]
Submitting mismatched passwords generates an error server side, but no message is displayed to the user.

**Comment by caseypt** [comment]
I was able to do the following:

- Successfully sign-up when all the fields are filled out correctly.
- Get an error when the username is already in use.

**Comment by caseypt** [comment]
If you don't fill in any of the fields, you get an error message for each field. However, if you fill in the username field, you then only get messages for the password fields. You should still get an error for the email field.

**Comment by caseypt** [comment]
Good start. I'll take another look once these issues are addressed.

**Comment by lewfish** [comment]
Thanks for all the feedback. I believe I've addressed all the concerns.

**Comment by lewfish** [comment]
I just pulled develop, rebased on top of develop, and there's still a commit from Matt in this PR. I'm not sure why.

**Comment by kdeloach** [comment]
I think you can safely remove that commit (bbdccd0c23a26b8e561dab69f1ecabcb3b4a704c) with interactive rebase.

**Comment by lewfish** [comment]
@caseypt Ok, this is good to go.

**Comment by lewfish** [comment]
@azavea-bot rebuild

**Comment by rajadain** [comment]
You may have to rebase over latest `develop` because I just merged #216 which touches some files that you also changed.

**Comment by caseypt** [comment]
Yeah, I get a few conflicts when pulling this down. When you fix those up, I'll take a look.

**Comment by lewfish** [comment]
Assuming it builds on Jenkins, this should be ready to test.

File: src/mmw/apps/user/views.py [hunk 1]
```diff
@@ -114,6 +114,57 @@ def itsi_register(request):
...
+    form = RegistrationFormUniqueEmail(request.POST)
+    if form.is_valid():
+        view.register(request, **form.cleaned_data)
+    else:
+        if 'username' not in form.cleaned_data:
+            response_data['username_valid'] = False
+        if 'password1' not in form.cleaned_data or \
+           'password2' not in form.cleaned_data or \
+           form.cleaned_data['password1'] != form.cleaned_data['password2']:
+            response_data['password_valid'] = False
```
Comment (caseypt) [inline src/mmw/apps/user/views.py]: Indentation is off here.

**Comment by caseypt** [comment]
Looks good. Tested signing up and retrieving a password. Verified that all of the bugs I reported previously have been fixed.

A few things to consider:

- I think we should redirect to the home page when logging out.
- If you click "Sign Up", then dismiss the modal, then click "Login", the Sign Up modal opens (and preserves and displays error messages if there were any) instead of the Login modal. I think clicking "Login" should always open the Login modal. Likewise, if you successfully sign up or fill out the lost password form, when you click "Login", you get the last modal from each of those workflows.
- You might be able to squash a few of these commits together.

**Comment by rajadain** [comment]
After I signed up, I got the "Confirm your email address" modal. After clicking that link in the email, I was taken to a new page that activated my new account. Back on the first page, the "Confirm your email address" modal stayed on. When I dismissed it and clicked "Login", the same "Confirm your email address" modal came up. But when I refreshed the page so that I could log in, I saw that I was already logged in as the new user.

To add to @caseypt's point above, I think that when clicking "Login" if the user has already been logged in then only that area should update to reflect that the user is logged in (just like clicking "Logout" does not refresh the page, simply updates that area). If the user is not already logged in, then they should see the Login modal.

**Comment by lewfish** [comment]
I updated according to the latest feedback.

**Comment by lewfish** [comment]
Sorry, I need to fix some unit tests first.

**Comment by lewfish** [comment]
To test the latest additions:

Sign up a new user and dismiss the modal, but don't go to the activation page. Click Login. This should show the login modal and should not say 'Continue as Guest'. 

Now activate the new account in a new tab, but don't click a link to login or go back to MMW. Switch back to MMW and click Login. It should switch from Login to the name of the user without having to go through the login modal.

**Comment by rajadain** [comment]
Good work. After dismissing the post-sign-up modal and clicking "Login" again, I was shown the Login modal and not the "Continue as Guest" modal. When I tried to login with the new credentials, I was shown the error "Invalid username or password", which I _guess_ is fine, although we could also have a message that says that the account hasn't been activated yet which is more meaningful for the user. How do others feel about this?

Also consider redirecting to the home page after account has been activated. Although I guess this could be potentially confusing if the user already has work in progress in the first tab.

**Comment by mmcfarland** [comment]
I agree for a new message about needing to activate the account.

For redirecting, we're supporting a workflow for anonymous users who have work in progress and decide to register at that point.  We should leave the register window completion modal up with a button for login available (or flip directly to the login) and in the activation tab, there should be instructions similar to "if you had work in progress, visit the previous tab to continue and save your work".  Activating an account should not log you in.

**Comment by mmcfarland** [comment]
re: my comments above, I don't think they need to be addressed in this PR.  We can incorporate them into a task about making the not-logged-in-save workflow functional.

**Comment by rajadain** [comment]
@lewfish to not auto-login a user when they activate their account I think you can change this one line to `False`: https://github.com/WikiWatershed/model-my-watershed/blob/develop/src/mmw/mmw/settings/base.py#L305

**Comment by caseypt** [comment]
+1. Tested. Works as described. Agree that other changes can come in a different issue.

If there isn't an issue consider creating one for the workflow described above as well as one for the custom modal transitions.

**Comment by lewfish** [comment]
Some new issues related to this PR:

#258 
#259 
#260

**Pull request merged**